README.md
# Dragon Scourge
- See also: [Dragon Knight](https://github.com/renderse7en/dragon-knight)
- [Live Demo](http://www.dragonscourge.com/play)

Many years ago, when I was young and dumb, I wrote a [simple little game](https://github.com/renderse7en/dragon-knight) based on the game *Dragon Warrior* for the NES. It was fun, it helped me learn how to code, and a lot of people liked it.

After that, I begin writing a sequel, of sorts, called Dragon Scourge. I never really finished that game. Life got in the way. Got married. Got a job. Had kids. All that.

I'm now turning it over to the open source community. 

As with Dragon Knight before it, please keep some things in mind:
- It's super old. It may not even work on modern versions of PHP. It may have security issues. I have no idea.
- I have moved on with my life, and am no longer changing or doing anything with this game.
- I am not providing help or support. You're on your own.
- I am not accepting pull requests. If you fork this, you are welcome to do whatever you want, but no changes will be merged back into this.
- Quite frankly, I don't really suggest that you use this as is. It's probably better as an inspiration for your own project. 
- Have fun with this. I gave it a lot of love a long time ago. I hope it inspires you to give something a lot of love as well.
- This Git repo represents the final released version, Beta 5 Build 20, originally released 9/2/2007.

# System Requirements
- PHP (4.1 and higher)
- MySQL
- zlib compression enabled on your server (optional)

# Installation Instructions
1. Clone this repo or download the zip.
2. Create a new database for Dragon Scourge to use, if you don't already have one set up.
3. Edit `config.php` to include the correct values for your database setup.
4. Upload the contents of the Dragon Scourge folder to your site.
5. Using a MySQL client such as HeidiSQL or PHPMyAdmin, run the contents of `install.sql` on the database you created.
7. Enjoy the game.

# License
MIT License

Copyright (c) 2017 renderse7en

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


changelog.txt
DRAGON SCOURGE
Changelog

***** BETA FIVE *****

9.2.2007 - Build 20 (Consolation Prize Part Deux):
- Fixed incorrect story values in install.sql.
- Some changes to the version/copyright/moddedby display.

07.28.2007 - Build 19 (Consolation Prize):
- Added "moddedby" column to controlrow. Will display next to normal Scourge copyright.
- Added SigBot, which lets players display their character stats in forum signature images.
- PVP fights now properly escape the opposing player's $userrow before updating.
- Fixed misspellings of 'bonusdefense' in spells.php.
- Fixed error that occurred when you failed to select a spell slot when adding new spells.
- Minor changes to err() output.

***** BETA FOUR *****

09.03.2006 - Build 18 (Joe Camel):
- Fixed function redeclaration with Blessed Strike & Stone Skin spells.
- When you're all alone, you no longer have a little blue "undefined" orb to keep you company.
- Changing worlds after a quest now gives you the world's starting town to avoid panel errors.
- Minor optimization fix for item building.
- Unique items no longer drop with prefixes/suffixes, and you can no longer buy unique items.
- Removed alphanumeric requirement for passwords.
- hpleech(monster) now properly accounts for difficulty multiplier when checking against the monster's maxhp.
- Added Deposit/Withdraw All links to the Bank.
- Fixed PVP and Post office bugs related to the new doquery syntax.
- Fixed single-quote bugs in mymail().
- Removed verbiage from the levelup template that was no longer necessary.
- Fixed some more verbiage in the registration message that was also no longer necessary.
- Added admin-definable cookie name and cookie domain to controlrow.

8.09.2006 - Build 17 (Haiku):
- Moved stylesheets into .css files.
- Added javascript tooltips for the class info when creating a new character.
- Changed doquery() format - rewrote all calls to this function to match the new format.
- Various changes to the structure of lib.php.
- Implemented Anman's increased attack and defense spells (Blessed Strike & Stone Skin).
- Fixed class display in profile view.
- Began admin control panel.
- Various little bitty fixes.
- You can no longer gamble 0 gold.
- Added new level stat type: life.
- Monsters can now do excellent hits, and you can now dodge monsters.
- Removed several columns from controlrow that are no longer used.

***** BETA THREE *****

4.26.2006 - Build 16 (Road Rage):
- Fixed a bug that prevented you from applying fewer than your max levelspell points.
- Added links back to the stores after buying maps/items or using the bank.
- Added a link back to town after gambling.
- Fixed the link in the error message when failing to create a guild.

4.08.2006 - Build 15 (Dead Man's Chest):
- Character profiles now display class name instead of ID numbers.
- Fixed incorrect minimum level display in guild start/apply error messages.
- Finally re-implemented gambling.
- Added very short tooltip descriptions of the classes when creating new characters.
- Spiffied up the Version Info page a little bit.
- A few other minor cosmetic changes.

2.26.2006 - Build 14 (Pete Rose):
- The poor sods who are stuck on dial-up can now select to not show the minimap in their account settings.
- Fixed a problem with guild names in paneltop.
- Fixed minor bug in account settings that caused a server error when not changing your password.
- Manually logging out of the game now resets your onlinetime so you no longer appear in Who's Online.
- Letting a duel time out after it's been accepted now counts against your duelling wins/losses.
- Overhauled dorow() to allow forcing array indexes by a specific column. Modified several functions accordingly.
- Guild Honor is now shown correctly on the main Guild Hall page immediately after an update.
- Guild shoutboxes now add the guild ID from userrow instead of in the GET request. More secure.
- Guild minimim join/start levels are now admin-editable in controlrow.
- updateuserrow() now array_maps addslashes to allow apostrophes in item names.

2.04.2006 - Build 13 (Unlucky):
- Dying now properly takes you back to the actual lat/lon of the world's first town (rather than 0,0 every time).
- Added 160+ new monsters for the later realms.
- Buy Maps now correctly only shows available maps from the Realm you're in.
- Fixed some weirdness with the Travel To menu in the second Realm.
- Added storyline/quests.
- Added item drops from monsters.
- Monsters now have a chance to swing first, and to block you from running.
- Added Quick Heal to panel_bottom.
- Users who are already a member of a guild no longer see the "Apply to Join" link on the Guild List page.
- Added Honor Points to guilds.
   * plus 1 point per member.
   * plus floored square root of total combined member experience.
   * plus 2 points per member PVP win.
   * minus 1 point per member PVP loss.
   * Calculated every 24 hours automatically, and upon every approve/remove/leave.
   * Displayed on the Guild List page.
- Added stats for your current items on the first Buy screen.
- You cannot create/join a guild until Level 10 to help prevent bank abuse.
- Added PVP logging - wins, losses, and highest lvl character killed.
- Hall of Fame now shows top 25 chars.
- Hall of Fame now uses stock mysqli_fetch_array() instead of custom dorow().

fight.php
<?php // fight.php :: primary fight controller.

//	Dragon Scourge
//
//	Program authors: Jamin Blount
//	Copyright (C) 2007 by renderse7en
//	Script Version 1.0 Beta 5 Build 20

//	You may not distribute this program in any manner, modified or
//	otherwise, without the express, written consent from
//	renderse7en.
//
//	You may make modifications, but only for your own use and
//	within the confines of the Dragon Scourge License Agreement
//	(see our website for that).

include("lib.php");
include("globals.php");
include("fightmods.php");

// Global userrow.
global $userrow;
if ($userrow["currentaction"] != "Fighting") { die(header("Location: index.php")); }

// Global monsterrow.
if($userrow["currentmonsterid"] != 0) {
    $monsterquery = doquery("SELECT * FROM monsters WHERE id='".$userrow["currentmonsterid"]."' LIMIT 1");
    $monsterrow = dorow($monsterquery);
} else {
    rollmonster();
}

// Now we get to the real guts of the game. Yippy skippy.
dofight();

function rollmonster() {
    
    global $userrow, $monsterrow, $fightrow;
    if($userrow["latitude"] < 0) { $latitude = $userrow["latitude"] * -1; } else { $latitude = $userrow["latitude"]; }
    if($userrow["longitude"] < 0) { $longitude = $userrow["longitude"] * -1; } else { $longitude = $userrow["longitude"]; }
    $maxlevel = ceil(max($latitude, $longitude) / 5);
    $minlevel = $maxlevel - 3;
    $monsterquery = doquery("SELECT * FROM monsters WHERE world='".$userrow["world"]."' AND level >= $minlevel AND level <= $maxlevel ORDER BY RAND() LIMIT 1");
    $monsterrow = dorow($monsterquery);
    
    $userrow["currentmonsterhp"] = (ceil(rand($monsterrow["maxhp"] * .75, $monsterrow["maxhp"]) * $userrow["difficulty"]));
    $userrow["currentmonsterid"] = $monsterrow["id"];
    $userrow["currentaction"] = "Fighting";
    
    $fightrow["track"] .= "init fight / roll monster\nmid: ".$monsterrow["id"]."\nmhp:".$userrow["currentmonsterhp"]."\n\n";
    
    updateuserrow();
    dofight();
    
}

function dofight() {
    
    global $userrow, $monsterrow, $fightrow, $spells;
    
    if (isset($_POST["fight"])) {
        
        $fightrow["track"] .= "init turn\nmid:".$userrow["currentmonsterid"]."\nmhp:".$userrow["currentmonsterhp"]."\n\n";
        
        playerturn();
        if ($userrow["currentmonsterhp"] <= 0) { youwin(); }
        monsterturn();
        if ($userrow["currenthp"] <= 0) { youlose(); }
        updateuserrow();
        
        $pagerow = array(
            "message"=>$fightrow["message"],
            "monstername"=>$monsterrow["name"],
            "monsterhp"=>$userrow["currentmonsterhp"],
            "playerphysdamage"=>$fightrow["playerphysdamage"],
            "playermagicdamage"=>$fightrow["playermagicdamage"],
            "playerfiredamage"=>$fightrow["playerfiredamage"],
            "playerlightdamage"=>$fightrow["playerlightdamage"],
            "monsterphysdamage"=>$fightrow["monsterphysdamage"],
            "monstermagicdamage"=>$fightrow["monstermagicdamage"],
            "monsterfiredamage"=>$fightrow["monsterfiredamage"],
            "monsterlightdamage"=>$fightrow["monsterlightdamage"],
            "track"=>$fightrow["track"]);
        $pagerow["spells"] = dospellslist();
        display("Fighting",parsetemplate(gettemplate("fight_turn"),$pagerow));
        
    } elseif (isset($_POST["spell"])) {
    
        if (! is_numeric($_POST["spellid"])) { err("Invalid spell selection."); }
        $isavailable = 0;
        for ($i = 1; $i < 11; $i++) {
            if ($userrow["spell".$i."id"] == $_POST["spellid"]) { $isavailable = 1; }
        }
        if ($isavailable == 0) { err("You don't have that spell."); }
        
        include("spells.php");
        $castspell = $spells[$_POST["spellid"]]["fname"]($_POST["spellid"]);
        $userrow["currentmonsterhp"] -= ($fightrow["playerphysdamage"] + $fightrow["playermagicdamage"] + $fightrow["playerfiredamage"] + $fightrow["playerlightdamage"]);
        if ($userrow["currentmonsterhp"] <= 0) { youwin(); }
        monsterturn();
        if ($userrow["currenthp"] <= 0) { youlose(); }
        updateuserrow();
        
        $pagerow = array(
            "message"=>$castspell,
            "monstername"=>$monsterrow["name"],
            "monsterhp"=>$userrow["currentmonsterhp"],
            "playerphysdamage"=>$fightrow["playerphysdamage"],
            "playermagicdamage"=>$fightrow["playermagicdamage"],
            "playerfiredamage"=>$fightrow["playerfiredamage"],
            "playerlightdamage"=>$fightrow["playerlightdamage"],
            "monsterphysdamage"=>$fightrow["monsterphysdamage"],
            "monstermagicdamage"=>$fightrow["monstermagicdamage"],
            "monsterfiredamage"=>$fightrow["monsterfiredamage"],
            "monsterlightdamage"=>$fightrow["monsterlightdamage"],
            "track"=>$fightrow["track"]);
        $pagerow["spells"] = dospellslist();
        display("Fighting",parsetemplate(gettemplate("fight_turn"),$pagerow));
    
    } elseif (isset($_POST["run"])) {
        
        if (rand(4,10) + ceil(sqrt($userrow["dexterity"])) < (rand(1,5) + ceil(sqrt((0.75 * $monsterrow["physattack"]))))) {
            
            monsterturn();
            $fightrow["message"] .= "You tried to run away, but the monster blocked you!<br />";
            if ($userrow["currenthp"] <= 0) { youlose(); }
            updateuserrow();
            
            $pagerow = array(
                "message"=>$fightrow["message"],
                "monstername"=>$monsterrow["name"],
                "monsterhp"=>$userrow["currentmonsterhp"],
                "monsterphysdamage"=>$fightrow["monsterphysdamage"],
                "monstermagicdamage"=>$fightrow["monstermagicdamage"],
                "monsterfiredamage"=>$fightrow["monsterfiredamage"],
                "monsterlightdamage"=>$fightrow["monsterlightdamage"],
                "track"=>$fightrow["track"]);
            $pagerow["spells"] = dospellslist();
            display("Fighting",parsetemplate(gettemplate("fight_monsteronly"),$pagerow));
            
        }
    
        $userrow["currentaction"] = "Exploring";
        $userrow["currentmonsterid"] = 0;
        $userrow["currentmonsterhp"] = 0;
        updateuserrow();
        die(header("Location: index.php"));
        
    } else {
        
        if (rand(1,10) + ceil(sqrt($userrow["dexterity"])) < (rand(1,7) + ceil(sqrt((0.75 * $monsterrow["physattack"]))))) {
 
            monsterturn();
            $fightrow["message"] .= "The monster attacked before you were ready!<br />";
            if ($userrow["currenthp"] <= 0) { youlose(); }
            updateuserrow();
            
            $pagerow = array(
                "message"=>$fightrow["message"],
                "monstername"=>$monsterrow["name"],
                "monsterhp"=>$userrow["currentmonsterhp"],
                "monsterphysdamage"=>$fightrow["monsterphysdamage"],
                "monstermagicdamage"=>$fightrow["monstermagicdamage"],
                "monsterfiredamage"=>$fightrow["monsterfiredamage"],
                "monsterlightdamage"=>$fightrow["monsterlightdamage"],
                "track"=>$fightrow["track"]);
            $pagerow["spells"] = dospellslist();
            display("Fighting",parsetemplate(gettemplate("fight_monsteronly"),$pagerow));
            
        }
    
        $pagerow = array(
            "monstername"=>$monsterrow["name"],
            "monsterhp"=>$userrow["currentmonsterhp"],
            "track"=>$fightrow["track"]);
        $pagerow["spells"] = dospellslist();
        display("Fighting",parsetemplate(gettemplate("fight_new"),$pagerow));
        
    }
    
}

function playerturn() {
    
    global $userrow, $monsterrow, $fightrow;
    
    // Calculate all damages.
    if ($userrow["physattack"] != 0) {
        $physhit = ceil(rand($userrow["physattack"]*.75, $userrow["physattack"]) / 3);
        $physblock = ceil(rand($monsterrow["physdefense"]*.75, $monsterrow["physdefense"]) / 3);
        $fightrow["playerphysdamage"] = max($physhit - $physblock, 1);
        $fightrow["track"] .= "physdam init: " . $fightrow["playerphysdamage"] . "\n";
    }
    
    if ($userrow["magicattack"] != 0) {
        $magichit = ceil(rand($userrow["magicattack"]*.75, $userrow["magicattack"]) / 3);
        $magicblock = ceil(rand($monsterrow["magicdefense"]*.75, $monsterrow["magicdefense"]) / 3);
        $fightrow["playermagicdamage"] = max($magichit - $magicblock, 0);
        $fightrow["track"] .= "magicdam init: " . $fightrow["playermagicdamage"] . "\n";
    }
    
    if ($userrow["fireattack"] != 0) {
        $firehit = ceil(rand($userrow["fireattack"]*.75, $userrow["fireattack"]) / 3);
        $fireblock = ceil(rand($monsterrow["firedefense"]*.75, $monsterrow["firedefense"]) / 3);
        $fightrow["playerfiredamage"] = max($firehit - $fireblock, 0);
        $fightrow["track"] .= "firedam init: " . $fightrow["playerfiredamage"] . "\n";
    }
    
    if ($userrow["lightattack"] != 0) {
        $lighthit = ceil(rand($userrow["lightattack"]*.75, $userrow["lightattack"]) / 3);
        $lightblock = ceil(rand($monsterrow["lightdefense"]*.75, $monsterrow["lightdefense"]) / 3);
        $fightrow["playerlightdamage"] = max($lighthit - $lightblock, 0);
        $fightrow["track"] .= "lightdam init: " . $fightrow["playerlightdamage"] . "\n";
    }
    
    // Chance to make an excellent hit.
    $toexcellent = rand(0,150);
    if ($toexcellent <= sqrt($userrow["dexterity"])) { 
        $fightrow["playerphysdamage"] *= 2;
        $fightrow["playermagicdamage"] *= 2;
        $fightrow["playerfiredamage"] *= 2;
        $fightrow["playerlightdamage"] *= 2;
        $fightrow["message"] .= "<b>Excellent hit!</b><br />"; 
        $fightrow["track"] .= "excellent\nphys:".$fightrow["playerphysdamage"]."\nmagic:".$fightrow["playermagicdamage"]."\nfire:".$fightrow["playerfiredamage"]."\nlight:".$fightrow["playerlightdamage"]."\n";
    }
    
    // Chance for monster to dodge.
    $tododge = rand(0,200);
    if ($tododge <= sqrt($monsterrow["physdefense"])) { 
        $fightrow["playerphysdamage"] = 0;
        $fightrow["playermagicdamage"] = 0;
        $fightrow["playerfiredamage"] = 0;
        $fightrow["playerlightdamage"] = 0;
        $fightrow["message"] .= "<b>The monster dodged your hit!</b><br />"; 
        $fightrow["track"] .= "monster dodge: all damage to zero\n";
    }
    
    // Now we add Per Turn mods.
    bonusattack();
    hpleech("player");
    mpleech("player");
    
    // Subtract all damage from monster's hp.
    $userrow["currentmonsterhp"] -= ($fightrow["playerphysdamage"] + $fightrow["playermagicdamage"] + $fightrow["playerfiredamage"] + $fightrow["playerlightdamage"]);
    $fightrow["track"] .= "end turn - mhp:".$userrow["currentmonsterhp"];
    
}

function monsterturn() {
    
    global $userrow, $monsterrow, $fightrow;
    
    if ($monsterrow["physattack"] != 0) {
        $physhit = ceil((rand($monsterrow["physattack"]*.75, $monsterrow["physattack"]) / 3) * $userrow["difficulty"]);
        $physblock = ceil(rand($userrow["physdefense"]*.75, $userrow["physdefense"]) / 3);
        $fightrow["monsterphysdamage"] = max($physhit - $physblock, 1); // Have to do at least 1 physical damage.
    }
    
    if ($monsterrow["magicattack"] != 0) {
        $magichit = ceil((rand($monsterrow["magicattack"]*.75, $monsterrow["magicattack"]) / 3) * $userrow["difficulty"]);
        $magicblock = ceil(rand($userrow["magicdefense"]*.75, $userrow["magicdefense"]) / 3);
        $fightrow["monstermagicdamage"] = max($magichit - $magicblock, 0);
    }
    
    if ($monsterrow["fireattack"] != 0) {
        $firehit = ((rand($monsterrow["fireattack"]*.75, $monsterrow["fireattack"]) / 3) * $userrow["difficulty"]);
        $fireblock = ceil(rand($userrow["firedefense"]*.75, $userrow["firedefense"]) / 3);
        $fightrow["monsterfiredamage"] = max($firehit - $fireblock, 0);
    }
    
    if ($monsterrow["lightattack"] != 0) {
        $lighthit = ceil((rand($monsterrow["lightattack"]*.75, $monsterrow["lightattack"]) / 3) * $userrow["difficulty"]);
        $lightblock = ceil(rand($userrow["lightdefense"]*.75, $userrow["lightdefense"]) / 3);
        $fightrow["monsterlightdamage"] = max($lighthit - $lightblock, 0);
    }
    
    // Chance to make an excellent hit.
    $toexcellent = rand(0,150);
    if ($toexcellent <= sqrt($monsterrow["physdefense"])) { 
        $fightrow["monsterphysdamage"] *= 2;
        $fightrow["monstermagicdamage"] *= 2;
        $fightrow["monsterfiredamage"] *= 2;
        $fightrow["monsterlightdamage"] *= 2;
        $fightrow["message"] .= "<b>The monster made an excellent hit!</b><br />"; 
    }
    
    // Chance for player to dodge.
    $tododge = rand(0,200);
    if ($tododge <= sqrt($userrow["physdefense"])) { 
        $fightrow["monsterphysdamage"] = 0;
        $fightrow["monstermagicdamage"] = 0;
        $fightrow["monsterfiredamage"] = 0;
        $fightrow["monsterlightdamage"] = 0;
        $fightrow["message"] .= "<b>You dodged the monster's hit!</b><br />"; 
    }
    
    // Now we add Per Turn mods.
    bonusdefense();
    hpleech("monster");
    
    // Subtract all damage from player's hp.
    $userrow["currenthp"] -= ($fightrow["monsterphysdamage"] + $fightrow["monstermagicdamage"] + $fightrow["monsterfiredamage"] + $fightrow["monsterlightdamage"]);
    
}

function youwin() {
    
    global $userrow, $monsterrow, $fightrow;
    
    $template = "fight_win";

    $newexp = ceil(rand($monsterrow["maxexp"]*.75, $monsterrow["maxexp"]) * (1 + ($userrow["expbonus"] / 100)));
    $newgold = ceil(rand($monsterrow["maxgold"]*.75, $monsterrow["maxgold"]) * (1 + ($userrow["goldbonus"] / 100)));
    $userrow["experience"] += $newexp;
    $userrow["gold"] += $newgold;
    $userrow["currentaction"] = "Exploring";
    $userrow["currentfight"] = 0;
    $userrow["currentmonsterid"] = 0;
    $userrow["currentmonsterhp"] = 0;
    if ($monsterrow["newstory"] != "0") {
        $userrow["story"] = $monsterrow["newstory"];
    }
    $userrow["bonusattack"] = 0;
    $userrow["bonusdefense"] = 0;
    
    // Now we add Per Kill mods.
    hpgain();
    mpgain();
    
    // Check for new levelup.
    if ($userrow["experience"] >= dolevels($userrow["level"]+1)) {
        $template = "fight_levelup";
        $userrow["level"]++;
        $userrow["levelup"] += 5;
        $userrow["maxtp"] += 5;
        $userrow["currenthp"] = $userrow["maxhp"];
        $userrow["currentmp"] = $userrow["maxmp"];
        $userrow["currenttp"] = $userrow["maxtp"];
        if (($userrow["level"] % 5 == 0)) { $userrow["levelspell"]++; $template = "fight_levelupspell"; }
    }
    
    // Roll for monster drop.
    if (rand(0,7) == 1) {
        
        // Grab lots of stuff from the DB.
        $preitemsrow = dorow(doquery("SELECT * FROM itembase WHERE reqlevel>='".($userrow["level"] - 5)."' AND reqlevel<='".$userrow["level"]."' AND willdrop='1' ORDER BY RAND() LIMIT 1", "itembase"));
        $preprefixrow = dorow(doquery("SELECT * FROM itemprefixes WHERE reqlevel<='".$userrow["level"]."' ORDER BY RAND() LIMIT 1", "itemprefixes"));
        $presuffixrow = dorow(doquery("SELECT * FROM itemsuffixes WHERE reqlevel<='".$userrow["level"]."' ORDER BY RAND() LIMIT 1", "itemsuffixes"));
        
        if ($preitemsrow["isunique"] == 1) {
            $idstring = "0," . $preitemsrow["id"] . ",0";
        } else {
            $idstring = "";
            if (rand(0,4)==1) { $idstring .= $preprefixrow["id"] . ","; } else { $idstring .= "0,"; }
            $idstring .= $preitemsrow["id"] . ",";
            if (rand(0,4)==1) { $idstring .= $presuffixrow["id"]; } else { $idstring .= "0"; }
        }
        $userrow["dropidstring"] = $idstring;
        $fightrow["message"] .= "<a href=\"index.php?do=itemdrop\" class=\"red\"><b>The monster has dropped an item! Click here for more information.</b></a><br />";
        
    }
    
    // Update for new stats.
    updateuserrow();
    
    // And we're done.
    $pagerow = array(
        "message"=>$fightrow["message"],
        "monstername"=>$monsterrow["name"],
        "monsterhp"=>$userrow["currentmonsterhp"],
        "playerphysdamage"=>$fightrow["playerphysdamage"],
        "playermagicdamage"=>$fightrow["playermagicdamage"],
        "playerfiredamage"=>$fightrow["playerfiredamage"],
        "playerlightdamage"=>$fightrow["playerlightdamage"],
        "monsterphysdamage"=>$fightrow["monsterphysdamage"],
        "monstermagicdamage"=>$fightrow["monstermagicdamage"],
        "monsterfiredamage"=>$fightrow["monsterfiredamage"],
        "monsterlightdamage"=>$fightrow["monsterlightdamage"],
        "newexp"=>$newexp,
        "newgold"=>$newgold);
    display("Victory!",parsetemplate(gettemplate($template),$pagerow));
    
}

function youlose() {
    
    global $userrow, $monsterrow, $fightrow;
    
    // First take away half the gold.
    $userrow["gold"] = ceil($userrow["gold"] / 2);
    
    // Then take away experience.
    $thislevel = dolevels($userrow["level"]);
    $nextlevel = dolevels($userrow["level"]+1);
    $userrow["experience"] -= ceil((($nextlevel - $thislevel) / 100) * $userrow["deathpenalty"]);
    
    // And we're done.
    $pagerow = array(
        "message"=>$fightrow["message"],
        "monstername"=>$monsterrow["name"],
        "monsterhp"=>$userrow["currentmonsterhp"],
        "playerphysdamage"=>$fightrow["playerphysdamage"],
        "playermagicdamage"=>$fightrow["playermagicdamage"],
        "playerfiredamage"=>$fightrow["playerfiredamage"],
        "playerlightdamage"=>$fightrow["playerlightdamage"],
        "monsterphysdamage"=>$fightrow["monsterphysdamage"],
        "monstermagicdamage"=>$fightrow["monstermagicdamage"],
        "monsterfiredamage"=>$fightrow["monsterfiredamage"],
        "monsterlightdamage"=>$fightrow["monsterlightdamage"],
        "deathpenalty"=>$userrow["deathpenalty"]);
        
    // Then put them in town & reset fight stuff.
    $townrow = dorow(doquery("SELECT * FROM towns WHERE world='".$userrow["world"]."' ORDER BY id ASC LIMIT 1"));
    $userrow["latitude"] = $townrow["latitude"];
    $userrow["longitude"] = $townrow["longitude"];
    $userrow["currentaction"] = "In Town";
    $userrow["currentfight"] = 0;
    $userrow["currentmonsterid"] = 0;
    $userrow["currentmonsterhp"] = 0;
    $userrow["currenthp"] = ceil($userrow["maxhp"] / 4);
    $userrow["bonusattack"] = 0;
    $userrow["bonusdefense"] = 0;
    
    // Update.
    updateuserrow();
    
    display("Thou Art Dead.",parsetemplate(gettemplate("fight_lose"),$pagerow));

}

function dolevels($level) {
    
    $leveltotal = 15;
    $leveladd = 15;
    $i = 2;
    while ($i < $level) {
        $levelstart = $leveltotal;
        if ($i < 4) { 
            $leveladd = ceil($leveladd * 2.0);
        } elseif ($i < 13) { 
            $leveladd = floor($leveladd * 1.45);
        } elseif ($i < 40) { 
            $leveladd = floor($leveladd * 1.20);
        } elseif ($i < 60) { 
            $leveladd = 150000;
        } elseif ($i < 80) { 
            $leveladd = 200000;
        } elseif ($i < 100) { 
            $leveladd = 300000;
        } elseif ($i >= 100) { 
            $leveladd = 500000;
        }
        $leveltotal = $levelstart + $leveladd;
        $i++;
    }
    return $leveltotal;
    
}

function dospellslist() {
    
    global $userrow, $spells;
    $options = "";
    for ($i = 1; $i < 11; $i++) {
        if ($userrow["spell".$i."id"] != 0) { 
            $options .= "<option value=\"".$userrow["spell".$i."id"]."\">".$userrow["spell".$i."name"]."</option>\n";
        }
    }
    if ($options != "") { 
        $list = "<select name=\"spellid\">$options</select> <input type=\"submit\" name=\"spell\" value=\"Cast Spell\" />";
    } else { $list = "<input type=\"submit\" name=\"spell\" value=\"Cast Spell\" disabled=\"disabled\" />"; }
    return $list;
    
}

?>
